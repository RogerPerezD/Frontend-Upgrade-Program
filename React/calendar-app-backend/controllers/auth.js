const { response } = require('express');
const User = require('../models/User');
const bcrypt = require('bcrypt');
const { generateJWT } = require('../helpers/jwt');

const createUser = async (req , resp = response)=>{
    const  { email, password}= req.body;

    try {
        let user = await User.findOne({email});
        if (user) {
            return resp.status(400).json({
                ok: false,
                msg: 'The email is already taken'
            });
        }
        user = new User( req.body );
        //Encriptar password
        const salt = bcrypt.genSaltSync();
        user.password = bcrypt.hashSync( password, salt );

        await user.save();

        // Generate JWT
        const token = await generateJWT( user.id, user.name);
    
        resp.status(201).json({
            ok: true,
            uid: user.id,
            name: user.name,
            token
        });
    } catch (error) {
        console.log(error);
        resp.status(500).json({
            ok: false,
            msg: 'Please contact the admin'
        });
    }

}

const loginUser = async(req, resp = response)=>{
    const  { email, password }= req.body;

    try {
        
        // Validate email
        let user = await User.findOne({email});
        if (!user) {
            return resp.status(400).json({
                ok: false,
                msg: 'The email doesnt exist'
            });
        }

        // Validate password
        const validPassword = bcrypt.compareSync( password, user.password );
        if (!validPassword) {
            return resp.status(400).json({
                ok: false,
                msg: 'The password is incorrect'
            });
        }

        // Generate JWT
        const token = await generateJWT( user.id, user.name);

        
        resp.status(200).json({
            ok: true,
            uid: user.id,
            name: user.name,
            token
        });
    } catch (error) {
        console.log(error);
        resp.status(500).json({
            ok: false,
            msg: 'Please contact the admin'
        });
    }
}

const revalidateToken = (req, resp = response)=>{
    resp.json({
        ok: true,
        msg: 'renewToken'
    });
}

module.exports = {
    createUser,
    loginUser,
    revalidateToken
}