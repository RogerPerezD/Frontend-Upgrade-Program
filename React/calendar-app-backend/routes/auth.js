/*
    User's routes / Auth
    host + /api/auth
*/

const { Router } = require('express');
const router = Router();
const { check } = require('express-validator');
const { createUser, loginUser, revalidateToken } = require('../controllers/AuthController');
const { validateFields } = require('../middlewares/validate-fields');
const { validateJWT } = require('../middlewares/validate-jwt');

router.post(
    '/new', 
    [//middlewares
        check('name', 'name is required').not().isEmpty(),
        check('email', 'email is required').isEmail(),
        check('password', 'password must have at least 6 characters').isLength({min: 6}),
        validateFields
    ],
    createUser);

router.post(
    '/', 
    [//middlewares
        check('email', 'email is required').isEmail(),
        check('password', 'password must have at least 6 characters').isLength({min: 6}),
        validateFields
    ],
    loginUser);

router.get(
    '/renew',
    [   //middlewares
        validateJWT
    ], 
    revalidateToken);

module.exports = router;