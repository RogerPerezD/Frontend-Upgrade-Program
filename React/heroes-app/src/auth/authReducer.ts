import { types } from "../types/types";

export type UserState = { logged: boolean, name?: string }

export type ActionReducer = {
    type: string,
    payload?: UserState
}

export const authReducer = ( state: UserState, action: ActionReducer):UserState =>{
    switch (action.type) {
        case types.login:
            return {
                ...action.payload,
                logged: true
            }
        case types.logout:
            return {
                logged: false
            };
        default:
            return state;
    }
}