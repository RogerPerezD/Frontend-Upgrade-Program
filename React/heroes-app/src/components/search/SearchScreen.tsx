
import { useForm } from '../../hooks/useForm';
import { FormEvent, useState } from 'react';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';
import { Heroes } from '../../data/heroes';

export const SearchScreen = () => {
    
    const [ {searchText}, handleInput] = useForm({
        searchText: ''
    });

    const [heroesFiltered, setHeroesFiltered] = useState<Heroes[]>( [] as Heroes[]);

    const handleSearch = (e: FormEvent<HTMLFormElement>) =>{
        e.preventDefault();
        setHeroesFiltered(getHeroesByName( searchText ));
    }

    return (
        <div>
            <h1>Search</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search</h4>
                    <hr />
                    <form onSubmit={ handleSearch }> 
                        <input 
                        type="text" 
                        placeholder='Search Hero'
                        className='form-control'
                        name='searchText'
                        autoComplete='off'
                        value={ searchText }
                        onChange={ handleInput }/>

                        <button
                        className='btn btn-outline-primary mt-2'
                        type='submit'>
                            Search...
                        </button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr />
                    {
                        heroesFiltered?.map(
                            (hero: any) => ( 
                                <HeroCard 
                                key={ hero.id }
                                { ...hero }/>
                            )
                        )
                    }
                </div>
            </div>
        </div>
        );
};
