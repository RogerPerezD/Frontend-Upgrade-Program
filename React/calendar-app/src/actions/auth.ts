import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch"
import { User, DispatchAuth } from '../reducers/authReducer';
import { types } from '../types/types';
import Swal from 'sweetalert2';

export const startLogin = ( email: string, password: string) =>{
    return async ( dispatch: DispatchAuth)=>{
        
        const resp = await fetchWithoutToken('auth', {email, password}, 'POST');
        const body = await resp.json();

        // console.log(body);
        if ( body.ok ) {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime().toString());
        
            dispatch( 
                login({
                _id: body.uid,
                name: body.name
                }) 
            );
        }else{
            Swal.fire('Error', body.msg, 'error');
        }

    }
}

const login = ( user: User) =>{
    return {
        type: types.authLogin,
        payload: user
    }
}

export const startRegister = ( email: string, password: string, name: string) =>{
    return async ( dispatch: DispatchAuth )=>{
        
        const resp = await fetchWithoutToken('auth/new', {email, password, name}, 'POST');
        const body = await resp.json();

        // console.log(body);
        if ( body.ok ) {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime().toString());
        
            dispatch( 
                register({
                _id: body.uid,
                name: body.name
                }) 
            );
        }else{
            Swal.fire('Error', body.msg, 'error');
        }

    }
}


const register = ( user: User) =>{
    return {
        type: types.authStartRegister,
        payload: user
    }
}

export const startChecking = ()=>{
    return async ( dispatch: DispatchAuth )=>{
        
        const resp = await fetchWithToken('auth/renew');
        const body = await resp.json();

        // console.log(body);
        if ( body.ok ) {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime().toString());
        
            dispatch( 
                login({
                _id: body.uid,
                name: body.name
                }) 
            );
        }else{
            dispatch(checkingFinish());
        }

    }
}

const checkingFinish = ()=>{
    return {
        type: types.authCheckingFinish
    }
}

export const startLogout = ()=>{
    return (dispatch: DispatchAuth) =>{
        // localStorage.removeItem('token');
        // localStorage.removeItem('token-init-date');
        localStorage.clear();
        dispatch( logout() );
    }
}

const logout = ()=>{
    return {
        type: types.authLogout
    }
}