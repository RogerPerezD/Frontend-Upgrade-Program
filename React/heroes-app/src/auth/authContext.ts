import { createContext } from "react";
import { ActionReducer, UserState } from "./authReducer";

type Auth = {
    user: UserState
    dispatch: React.Dispatch<ActionReducer>
}

export const AuthContext = createContext( {} as Auth);