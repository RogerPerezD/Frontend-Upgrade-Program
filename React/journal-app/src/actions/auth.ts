import Swal from 'sweetalert2';
import { types } from '../types/types';
import { UserAction, DispatchTypeUser } from '../reducers/authReducer';
import { googleAuthProvider,firebase } from '../firebase/firebaseConfig';
import { DispatchTypeUI } from '../reducers/uiReducer';
import { finishLoading, removeError, setError, startLoading } from './ui';
import { DispatchTypeNote } from '../reducers/notesReducer';
import { noteLogout } from './notes';

export const startLoginEmailPassword = ( email: string, password: string) =>{
    return (( dispatch: (DispatchTypeUI & DispatchTypeUser) ) =>{

        dispatch( startLoading() );

        return firebase.auth().signInWithEmailAndPassword( email, password)
            .then(({user}) => {
                dispatch( login( (user?.uid as string), (user?.displayName as string)));
                dispatch( removeError() );
                dispatch( finishLoading() );
            })
            .catch( (e: Error) =>{
                dispatch( setError(e.message) );
                dispatch( finishLoading() );
                Swal.fire({
                    title: 'Error!',
                    text: e.message,
                    icon: 'error',
                    confirmButtonText: 'Confirm'
                });
            });
    });
}

export const startGoogleLogin = () =>{
    return ( dispatch: DispatchTypeUser ) =>{
        firebase.auth().signInWithPopup( googleAuthProvider )
        .then( ({user}) => {
            dispatch( login( (user?.uid as string), (user?.displayName as string)));
        });
    }
}

export const login = ( uid: string, displayName: string) =>{ 
    const action: UserAction = {
        type: types.login,
        payload: {
            uid,
            name: displayName
        }
    }
    return action;
}

export const  startRegisterWhitEmailAndPassword = ( email: string, password: string, name: string)=>{

    return ( dispatch: DispatchTypeUser ) => {
        firebase.auth().createUserWithEmailAndPassword( email, password )
        .then( async ({user}) => {
            // Put user's name in firebase 
            await user?.updateProfile({ displayName: name });

            // After register, the user is logged in.
            dispatch( login( (user?.uid as string), (user?.displayName as string)));
        });
    }
}

export const startLogout = () =>{
    return async ( dispatch: (DispatchTypeUser & DispatchTypeNote) )=>{
        await firebase.auth().signOut();
        dispatch( logout() );
        dispatch( noteLogout() );
    }
}

export const logout = (): UserAction =>{
    return {
        type: types.logout
    }
}