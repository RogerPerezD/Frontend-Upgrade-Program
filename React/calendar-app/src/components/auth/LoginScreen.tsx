import React from 'react';
import '../../css/login.css';
import { useForm } from '../../hooks/useForm';
import { FormEvent } from 'react';
import { useDispatch } from 'react-redux';
import { startLogin, startRegister } from '../../actions/auth';
import Swal from 'sweetalert2';

type LoginForm = {
    lEmail: string,
    lPassword: string
}

type RegisterForm = {
    rEmail: string,
    rPassword: string,
    rPasswordVerified: string,
    rName: string,
}

export const LoginScreen = () => {

    const dispatch = useDispatch();

    const [ formLoginValues, handleLoginInputChange ]= useForm<LoginForm>({
        lEmail: 'rogelio@mail.com',
        lPassword: '1234567'
    });

    const [ formRegisterValues, handleRegisterInputChange ]= useForm<RegisterForm>({
        rEmail: 'raul@mail.com',
        rPassword: '123456',
        rPasswordVerified: '123456',
        rName: 'Raul'
    });

    const { lEmail, lPassword } = formLoginValues;

    const { rName, rEmail, rPassword, rPasswordVerified} = formRegisterValues;

    const handleLogin = ( e: FormEvent)=>{
        e.preventDefault();
        console.log(lEmail, lPassword,'valores login');
        dispatch( startLogin(lEmail, lPassword) );
    }

    const handleRegister = ( e: FormEvent)=>{
        e.preventDefault();
        if (rPassword !== rPasswordVerified) {
            return Swal.fire('Error', 'Passwords must match','error')
        }
        dispatch( startRegister(rEmail, rPassword, rName) );
    }
    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Ingreso</h3>
                    <form onSubmit={ handleLogin }>
                        <div className="form-group">
                            <input 
                                type="text"
                                className="form-control"
                                placeholder="Correo"
                                name='lEmail'
                                value={lEmail}
                                onChange= {handleLoginInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contraseña"
                                name='lPassword'
                                value={lPassword}
                                onChange= {handleLoginInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input 
                                type="submit"
                                className="btnSubmit"
                                value="Login" 
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Registro</h3>
                    <form onSubmit={ handleRegister }>
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Nombre"
                                name='rName'
                                value={rName}
                                onChange= { handleRegisterInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Correo"
                                name='rEmail'
                                value={rEmail}
                                onChange= { handleRegisterInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Contraseña"
                                name='rPassword' 
                                value={rPassword}
                                onChange= { handleRegisterInputChange }
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Repita la contraseña"
                                name='rPasswordVerified'
                                value={rPasswordVerified}
                                onChange= { handleRegisterInputChange } 
                            />
                        </div>

                        <div className="form-group">
                            <input 
                                type="submit" 
                                className="btnSubmit" 
                                value="Crear cuenta" />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}