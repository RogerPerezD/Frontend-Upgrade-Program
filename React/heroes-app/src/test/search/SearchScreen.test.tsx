import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { AuthContext } from "../../auth/authContext";
import { SearchScreen } from "../../components/search/SearchScreen";

describe('Test in <SearchScreen/>', () => {

    test('should show the component whit default values', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen/>
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();

        expect( wrapper.find('.alert').text().trim()).toBe('Search a hero')
    });


    
});
