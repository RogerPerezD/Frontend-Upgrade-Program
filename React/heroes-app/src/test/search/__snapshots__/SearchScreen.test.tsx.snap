// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <SearchScreen/> should show the component whit default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search Hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-2"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
